File: core/src/rng/rng.ts
export interface RngState {
  seed: string;
  state: number;
  calls: number;
}

export interface WeightedEntry<T> {
  value: T;
  weight: number;
}

function hashSeed(seed: number | string): number {
  const text = String(seed);
  let hash = 2166136261;
  for (let index = 0; index < text.length; index += 1) {
    hash ^= text.charCodeAt(index);
    hash = Math.imul(hash, 16777619);
  }
  return (hash >>> 0) || 0x9e3779b9;
}

function nextUint32(state: number): number {
  let value = state >>> 0;
  value ^= value << 13;
  value ^= value >>> 17;
  value ^= value << 5;
  value >>>= 0;
  return value || 0x6d2b79f5;
}

export function createRngState(seed: number | string): RngState {
  return {
    seed: String(seed),
    state: hashSeed(seed),
    calls: 0
  };
}

export function nextFloat(rng: RngState): { value: number; rng: RngState } {
  const nextState = nextUint32(rng.state);
  return {
    value: nextState / 0x1_0000_0000,
    rng: {
      seed: rng.seed,
      state: nextState,
      calls: rng.calls + 1
    }
  };
}

export function nextInt(
  rng: RngState,
  minInclusive: number,
  maxExclusive: number
): { value: number; rng: RngState } {
  if (!Number.isInteger(minInclusive) || !Number.isInteger(maxExclusive)) {
    throw new Error("nextInt requires integer bounds.");
  }
  if (maxExclusive <= minInclusive) {
    throw new Error(
      `nextInt requires maxExclusive (${maxExclusive}) to be greater than minInclusive (${minInclusive}).`
    );
  }

  const span = maxExclusive - minInclusive;
  const { value, rng: nextRng } = nextFloat(rng);
  const sampled = minInclusive + Math.floor(value * span);
  return { value: sampled, rng: nextRng };
}

export function pickWeighted<T>(
  rng: RngState,
  entries: readonly WeightedEntry<T>[]
): { value: T; rng: RngState } {
  const validEntries = entries.filter((entry) => entry.weight > 0);
  if (validEntries.length === 0) {
    throw new Error("pickWeighted requires at least one entry with positive weight.");
  }

  const totalWeight = validEntries.reduce((sum, entry) => sum + entry.weight, 0);
  const { value: roll, rng: nextRng } = nextFloat(rng);
  let cursor = roll * totalWeight;
  for (const entry of validEntries) {
    if (cursor < entry.weight) {
      return { value: entry.value, rng: nextRng };
    }
    cursor -= entry.weight;
  }

  return { value: validEntries[validEntries.length - 1]!.value, rng: nextRng };
}

File: bit_life_survival/core/rng.py
from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Generic, Sequence, TypeVar

T = TypeVar("T")


def seed_to_uint32(seed: int | str) -> int:
    text = str(seed).encode("utf-8")
    digest = hashlib.sha256(text).hexdigest()
    value = int(digest[:8], 16)
    return value if value != 0 else 0x9E3779B9


@dataclass(slots=True)
class WeightedEntry(Generic[T]):
    value: T
    weight: float


@dataclass(slots=True)
class DeterministicRNG:
    seed: int | str
    state: int
    calls: int = 0

    @classmethod
    def from_seed(cls, seed: int | str) -> "DeterministicRNG":
        return cls(seed=seed, state=seed_to_uint32(seed), calls=0)

    def _next_uint32(self) -> int:
        value = self.state & 0xFFFFFFFF
        value ^= (value << 13) & 0xFFFFFFFF
        value ^= (value >> 17) & 0xFFFFFFFF
        value ^= (value << 5) & 0xFFFFFFFF
        value &= 0xFFFFFFFF
        self.state = value if value != 0 else 0x6D2B79F5
        self.calls += 1
        return self.state

    def next_float(self) -> float:
        return self._next_uint32() / 2**32

    def next_int(self, min_inclusive: int, max_exclusive: int) -> int:
        if max_exclusive <= min_inclusive:
            raise ValueError(
                f"next_int requires max_exclusive ({max_exclusive}) > min_inclusive ({min_inclusive})."
            )
        span = max_exclusive - min_inclusive
        return min_inclusive + int(self.next_float() * span)

    def pick_weighted(self, entries: Sequence[WeightedEntry[T]]) -> T:
        valid_entries = [entry for entry in entries if entry.weight > 0]
        if not valid_entries:
            raise ValueError("pick_weighted requires at least one positive weight.")

        total_weight = sum(entry.weight for entry in valid_entries)
        cursor = self.next_float() * total_weight
        for entry in valid_entries:
            if cursor < entry.weight:
                return entry.value
            cursor -= entry.weight
        return valid_entries[-1].value
